### PR TITLE
feat(flags): fall back to `/decide` endpoint for `GetFeatureFlag` and `GetAllFlags` so that users can use this library without needing a personal API Key

### PR DESCRIPTION
**Pull request by dmarticus**
Does what it says on the tin.  I made the following changes:

- Fallback logic in `GetFeatureFlag` and `GetAllFlags` that falls back to using the `/decide` endpoint for fetching feature flags if there's no PersonalApiKey provided
- Tests for this new behavior

Potential TODOs:
- Log the fallback?
- improve error messages with missing token to tell the users they can fall back?

Other things:
- Doc update associated with these changes: https://github.com/PostHog/posthog.com/pull/9121

### CODE REVIEW

File: feature_flags_test.go [hunk 1]
```diff
@@ -325,7 +325,7 @@ func TestFlagGroup(t *testing.T) {
 				t.Errorf("Expected personProperties to be map[region:Canada], got %s", reqBody.PersonProperties)
 			}
 
-			groupPropertiesEquality := reflect.DeepEqual(reqBody.GroupProperties, map[string]Properties{"company": Properties{"name": "Project Name 1"}})
+			groupPropertiesEquality := reflect.DeepEqual(reqBody.GroupProperties, map[string]Properties{"company": {"name": "Project Name 1"}})
```
Comment (dmarticus) [inline feature_flags_test.go:328]: compiler told me that `Properties` field was unused.

File: featureflags.go [hunk 2]
```diff
@@ -311,7 +311,7 @@ func (poller *FeatureFlagsPoller) computeFlagLocally(
 		groupName, exists := poller.groups[fmt.Sprintf("%d", *flag.Filters.AggregationGroupTypeIndex)]
 
 		if !exists {
-			errMessage := "Flag has unknown group type index"
+			errMessage := "flag has unknown group type index"
```
Comment (dmarticus) [inline featureflags.go:314]: compiler told me to lowercase error messages

File: featureflags.go [hunk 3]
```diff
@@ -467,7 +467,7 @@ func matchCohort(property FlagProperty, properties Properties, cohorts map[strin
 	cohortId := fmt.Sprint(property.Value)
 	propertyGroup, ok := cohorts[cohortId]
 	if !ok {
-		return false, fmt.Errorf("Can't match cohort: cohort %s not found", cohortId)
+		return false, fmt.Errorf("can't match cohort: cohort %s not found", cohortId)
```
Comment (dmarticus) [inline featureflags.go:470]: compiler told me to lowercase error messages

File: featureflags.go [hunk 4]
```diff
@@ -578,7 +578,7 @@ func matchProperty(property FlagProperty, properties Properties) (bool, error) {
 		return false, &InconclusiveMatchError{"Can't match properties with operator is_not_set"}
 	}
 
-	override_value, _ := properties[key]
+	override_value := properties[key]
```
Comment (dmarticus) [inline featureflags.go:581]: compiler told me this was unused.

File: featureflags.go [hunk 5]
```diff
@@ -637,7 +637,7 @@ func matchProperty(property FlagProperty, properties Properties) (bool, error) {
 			valueString = strconv.Itoa(valueInt)
 			r, err = regexp.Compile(valueString)
 		} else {
-			errMessage := "Regex expression not allowed"
+			errMessage := "regex expression not allowed"
```
Comment (dmarticus) [inline featureflags.go:640]: compiler told me to lowercase error messages

File: featureflags.go [hunk 6]
```diff
@@ -653,7 +653,7 @@ func matchProperty(property FlagProperty, properties Properties) (bool, error) {
 			valueString = strconv.Itoa(valueInt)
 			match = r.MatchString(valueString)
 		} else {
-			errMessage := "Value type not supported"
+			errMessage := "value type not supported"
```
Comment (dmarticus) [inline featureflags.go:656]: compiler told me to lowercase error messages

File: featureflags.go [hunk 7]
```diff
@@ -707,12 +707,12 @@ func matchProperty(property FlagProperty, properties Properties) (bool, error) {
 func validateOrderable(firstValue interface{}, secondValue interface{}) (float64, float64, error) {
 	convertedFirstValue, err := interfaceToFloat(firstValue)
 	if err != nil {
-		errMessage := "Value 1 is not orderable"
+		errMessage := "value 1 is not orderable"
```
Comment (dmarticus) [inline featureflags.go:710]: compiler told me to lowercase error messages

File: featureflags.go [hunk 8]
```diff
@@ -707,12 +707,12 @@ func matchProperty(property FlagProperty, properties Properties) (bool, error) {
...
 		return 0, 0, errors.New(errMessage)
 	}
 	convertedSecondValue, err := interfaceToFloat(secondValue)
 	if err != nil {
-		errMessage := "Value 2 is not orderable"
+		errMessage := "value 2 is not orderable"
```
Comment (dmarticus) [inline featureflags.go:715]: compiler told me to lowercase error messages

File: featureflags.go [hunk 9]
```diff
@@ -809,7 +809,7 @@ func (poller *FeatureFlagsPoller) GetFeatureFlags() ([]FeatureFlag, error) {
 	_, closed := <-poller.loaded
 	if closed && poller.featureFlags == nil {
 		// There was an error with initial flag fetching
-		return nil, fmt.Errorf("Flags were not successfully fetched yet")
+		return nil, fmt.Errorf("flags were not successfully fetched yet")
```
Comment (dmarticus) [inline featureflags.go:812]: compiler told me to lowercase error messages

File: posthog.go [hunk 10]
```diff
@@ -238,17 +250,28 @@ func (c *client) Enqueue(msg Message) (err error) {
...
+func (c *client) GetLastCapturedEvent() *Capture {
+	c.lastEventMutex.RLock()
+	defer c.lastEventMutex.RUnlock()
+	if c.lastCapturedEvent == nil {
+		return nil
+	}
+	// Return a copy to avoid data races
+	eventCopy := *c.lastCapturedEvent
+	return &eventCopy
+}
```
Comment (dmarticus) [inline posthog.go:268]: this is a handy method for tracking captured events, but I mostly only used it for testing.  Perhaps I shouldn't expose it on the API client?

File: posthog.go [hunk 11]
```diff
@@ -238,17 +250,28 @@ func (c *client) Enqueue(msg Message) (err error) {
...
 func (c *client) IsFeatureEnabled(flagConfig FeatureFlagPayload) (interface{}, error) {
 	if err := flagConfig.validate(); err != nil {
 		return false, err
 	}
 
-	if c.featureFlagsPoller == nil {
-		errorMessage := "specifying a PersonalApiKey is required for using feature flags"
-		c.Errorf(errorMessage)
-		return false, errors.New(errorMessage)
-	}
```
Comment (dmarticus) [inline posthog.go:250]: this method falls back to `GetFeatureFlag`, which in turn falls back, so I don't need this guard clause.

File: posthog.go [hunk 12]
```diff
@@ -314,12 +343,16 @@ func (c *client) GetAllFlags(flagConfig FeatureFlagPayloadNoKey) (map[string]int
...
-		errorMessage := "specifying a PersonalApiKey is required for using feature flags"
-		c.Errorf(errorMessage)
-		return nil, errors.New(errorMessage)
+	var flagsValue map[string]interface{}
+	var err error
+
+	if c.featureFlagsPoller != nil {
+		flagsValue, err = c.featureFlagsPoller.GetAllFlags(flagConfig)
+	} else {
+		flagsValue, err = c.getAllFeatureFlagsFromDecide(flagConfig.DistinctId, flagConfig.Groups, flagConfig.PersonProperties, flagConfig.GroupProperties)
```
Comment (dmarticus) [inline posthog.go:352]: didn't add the comments here that the above method has, I'll add them back in for cleanliness

File: posthog.go [hunk 13]
```diff
@@ -336,7 +369,7 @@ func (c *client) Close() (err error) {
 	return
 }
 
-// Asychronously send a batched requests.
+// Asynchronously send a batched requests.
```
Comment (dmarticus) [inline posthog.go:372]: drive-by spellcheck

File: posthog_test.go [hunk 14]
```diff
@@ -705,7 +706,7 @@ func TestClientMaxConcurrentRequests(t *testing.T) {
 			func(m APIMessage, e error) { errchan <- e },
 		},
 		Transport: testTransportDelayed,
-		// Only one concurreny request can be submitted, because the transport
+		// Only one concurrency request can be submitted, because the transport
```
Comment (dmarticus) [inline posthog_test.go:709]: drive-by spellcheck

File: posthog_test.go [hunk 15]
```diff
@@ -741,21 +742,9 @@ func TestFeatureFlagsWithNoPersonalApiKey(t *testing.T) {
...
-		},
-	)
-	_, receivedErrors[2] = client.GetFeatureFlag(
-		FeatureFlagPayload{
-			Key:        "some key",
-			DistinctId: "some id",
-		},
-	)
-	_, receivedErrors[3] = client.GetFeatureFlags()
+	_, receivedErrors[1] = client.GetFeatureFlags()
```
Comment (dmarticus) [inline posthog_test.go:747]: we only have 2 methods that error on a missing PersonalApiKey, and neither of them are hot paths (i.e. users don't need them to use our feature flags).

File: posthog_test.go [hunk 16]
```diff
@@ -766,6 +755,442 @@ func TestFeatureFlagsWithNoPersonalApiKey(t *testing.T) {
 
 }
 
+func TestIsFeatureEnabled(t *testing.T) {
```
Comment (dmarticus) [inline posthog_test.go:758]: this method didn't have tests, so now it does. Boy scout principle!

File: posthog_test.go [hunk 17]
```diff
@@ -766,6 +755,442 @@ func TestFeatureFlagsWithNoPersonalApiKey(t *testing.T) {
...
+				}
+				if result != tt.expectedResult {
+					t.Errorf("Expected result %v, got %v", tt.expectedResult, result)
+				}
+			}
+		})
+	}
+}
+
+func TestGetFeatureFlagWithNoPersonalApiKey(t *testing.T) {
```
Comment (dmarticus) [inline posthog_test.go:856]: test the fallbacks

File: posthog_test.go [hunk 18]
```diff
@@ -788,11 +1213,6 @@ func TestSimpleFlagOld(t *testing.T) {
 	if checkErr != nil || isEnabled != true {
 		t.Errorf("simple flag with null rollout percentage should be on for everyone")
 	}
-
-	// flagValue, valueError := client.GetFeatureFlag("simpleFlag", "hey", false, Groups{}, NewProperties(), map[string]Properties{})
-	// if valueError != nil || flagValue != true {
-	// 	t.Errorf("simple flag with null rollout percentage should have value 'true'")
-	// }
```
Comment (dmarticus) [inline posthog_test.go:795]: this was dead code

File: posthog.go [hunk 19]
```diff
@@ -278,12 +301,17 @@ func (c *client) GetFeatureFlag(flagConfig FeatureFlagPayload) (interface{}, err
...
-		errorMessage := "specifying a PersonalApiKey is required for using feature flags"
-		c.Errorf(errorMessage)
-		return "false", errors.New(errorMessage)
+	var flagValue interface{}
+	var err error
+
+	if c.featureFlagsPoller != nil {
+		// get feature flags from the poller, which uses the personal api key
+		flagValue, err = c.featureFlagsPoller.GetFeatureFlag(flagConfig)
+	} else {
```
Comment (dmarticus) [inline posthog.go:310]: should I log that we're falling back here?  Does it matter?

File: posthog.go (hunk 10 above)
Comment (fuziontech) [reply posthog.go:268]: eh, should be fine. Doesn't seem too unsafe to expose

File: posthog.go (hunk 19 above)
Comment (fuziontech) [reply posthog.go:310]: maybe, but only at a trace/info/debug level. Might be good for investigations.

**Review by fuziontech — APPROVED** [review]
image

A few comments, but otherwise this is sick
🚢 it

PS nice tests!

**Review by Phanatic — APPROVED** [review]
Ship it!
